llvm/include/llvm/Transforms/Utils/Local.h

```diff
@@ -556,6 +556,27 @@ Value *invertCondition(Value *Condition);
 /// function, explicitly materialize the maximal set in the IR.
 bool inferAttributesFromOthers(Function &F);
 
+//===----------------------------------------------------------------------===//
+//  Helpers to track and update flags on instructions.
+//
+
+struct OverflowTracking {
+  bool HasNUW = true;
+  bool HasNSW = true;
+  bool AllKnownNonNegative = true;
```

Comment:
Please add a comment indicating that the user is responsible for updating `AllKnownNonNegative` and `AllKnownNonZero`.

Reply:
Done thanks